File: web/src/types/quant.ts
export type CapabilityState = 'available' | 'configured' | 'unavailable' | 'degraded' | 'data_dependent';
export type QuantMarket = 'US' | 'CN';

export interface QuantCapabilities {
  status: CapabilityState;
  market: QuantMarket;
  pythonVersion: string;
  dailyPriceSemantics: 'forward_adjusted';
  markets: Record<string, CapabilityState>;
  qlib: { status: CapabilityState; version: string; execution: string; reason: string | null };
  models: { status: CapabilityState; required: Record<string, 'production' | 'unavailable'> };
  warnings: string[];
}

export interface QuantUniverseMember {
  code: string;
  name: string | null;
  sectorKey: string | null;
  sectorBenchmarkCode: string | null;
  effectiveFrom: string | null;
  effectiveTo: string | null;
}
export interface QuantUniverse {
  id: number;
  key: string;
  name: string;
  market: QuantMarket;
  benchmarkCode: string | null;
  config: Record<string, unknown>;
  memberCount: number;
  members: QuantUniverseMember[];
}
export interface QuantModelDefinition {
  id: number;
  key: string;
  name: string;
  modelType: string;
  taskType: string;
  frequency: string;
  enabled: boolean;
  supportedMarkets: QuantMarket[];
}
export interface QuantDatasetSnapshot {
  id: number;
  datasetKey: string;
  market: QuantMarket;
  dateFrom: string;
  dateTo: string;
  featureVersion: string;
  artifactUri: string | null;
  rowCount: number;
  symbolCount: number;
  universeMemberCount: number;
  universeCoverageRatio: number;
  minimumUniverseCoverage: number;
  trainable: boolean;
  status: 'pending' | 'building' | 'ready' | 'failed';
  validationResult: Record<string, unknown>;
  createdAt: string;
  finishedAt: string | null;
}
export interface DatasetBuildAccepted { taskId: string; status: string; market: QuantMarket; universe: string }
export interface ModelRunCreateAccepted { modelRunId: number; taskId: string; status: 'pending'; market: QuantMarket }
export interface QuantDeleteResult { id: number; deleted: boolean; artifactDeleted: boolean }
export interface ModelRunCreatePayload {
  market: QuantMarket;
  modelKey: 'cross_section_lgbm' | 'time_series_lgbm';
  modelVersion: string;
  datasetSnapshotId: number;
}
export interface MarketScoreComponent {
  key: string;
  label: string;
  group: string;
  groupLabel: string;
  rawValue: number;
  rawFormat: 'percent' | 'number';
  score: number;
  weight: number;
  contribution: number;
}
export interface MarketScoreGroup {
  key: string;
  label: string;
  weight: number;
  score: number;
  contribution: number;
  components: MarketScoreComponent[];
}
export interface MarketScoreBreakdown {
  version: string;
  score: number;
  weightTotal: number;
  groups: MarketScoreGroup[];
}
export interface MarketRegime {
  id: number;
  tradeDate: string;
  market: string;
  modelVersion: string;
  regime: 'risk_on'|'neutral'|'risk_off';
  marketScore: number;
  maxEquityExposure: number;
  features: Record<string, unknown>;
  reasons: string[];
  scoreBreakdown: MarketScoreBreakdown | null;
}
export interface SectorRegime { market: QuantMarket; tradeDate: string; sectorKey: string; benchmarkCode: string; benchmarkName: string|null; sectorScore: number; rank: number; state: string; features: Record<string, number|null> }
export interface QuantSignal { id: number; tradeDate: string; market: QuantMarket; code: string; name:string|null; modelVersion: string; finalScore: number; rawFinalScore: number; gatedFinalScore: number; marketScore: number|null; sectorScore: number|null; timeSeriesScore: number|null; crossSectionScore: number|null; riskPenalty: number; universeRank: number|null; sectorRank: number|null; predictedReturn: number|null; signal: string; targetPosition: number; vetoed: boolean; vetoReason: string|null; reasons: string[]; scoreComponents: Record<string, number|null> }
export interface SignalRanking { tradeDate: string|null; market: QuantMarket; universe: string; modelVersion: string|null; marketRegime: string|null; maxEquityExposure: number|null; items: QuantSignal[] }
export interface ModelRun { id:number; modelKey:string; modelVersion:string; runType:string; market:string; status:string; progress:number; trainStart:string|null; trainEnd:string|null; validStart:string|null; validEnd:string|null; testStart:string|null; testEnd:string|null; metrics:Record<string,number|null>; warnings:string[]; error:string|null; artifactUri:string|null; createdAt:string }
export interface PortfolioItem { id:number; code:string; name:string|null; sectorKey:string|null; rank:number; previousRank:number|null; action:string; currentWeight:number; targetWeight:number; weightChange:number; finalScore:number; predictedReturn:number|null; reasons:string[]; constraints:Record<string,unknown> }
export interface Portfolio { id:number; tradeDate:string; market:QuantMarket; universe:string; modelVersion:string; status:string; maxEquityExposure:number; targetEquityExposure:number; summary:Record<string,unknown>; warnings:string[]; generatedAt:string; items:PortfolioItem[] }
export interface IntradayConfirmation { id:number; tradeDate:string; code:string; name:string|null; evaluatedAt:string; decision:'confirm'|'wait'|'reject'|'expired'|'insufficient_data'; confidence:number; price:number|null; vwap:number|null; priceVsVwap:number|null; vwapSlope:number|null; first30mReturn:number|null; volumeRatio:number|null; relativeStrengthMarket:number|null; relativeStrengthSector:number|null; reasons:string[] }

File: src/finance_analysis/database/models/quant.py
"""Persistence models for versioned quantitative research results.

PostgreSQL remains the source of truth.  Qlib datasets and model binaries are
referenced by URI and never stored in JSON columns.
"""

from __future__ import annotations

from finance_analysis.core.time import utc_now
from finance_analysis.database.base import Base
from sqlalchemy import (
    BigInteger,
    Boolean,
    CheckConstraint,
    Column,
    Date,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship


def json_object():
    return {}


def json_array():
    return []


class QuantDatasetSnapshot(Base):
    __tablename__ = "quant_dataset_snapshot"
    id = Column(BigInteger, primary_key=True)
    dataset_key = Column(String(128), nullable=False, unique=True)
    market = Column(String(8), nullable=False)
    universe_id = Column(Integer, ForeignKey("universe.id", ondelete="RESTRICT"), nullable=False)
    frequency = Column(String(16), nullable=False)
    date_from = Column(Date, nullable=False)
    date_to = Column(Date, nullable=False)
    feature_version = Column(String(64), nullable=False)
    source_revision = Column(String(64), nullable=False)
    code_commit = Column(String(64))
    artifact_uri = Column(Text)
    row_count = Column(BigInteger, nullable=False, default=0)
    symbol_count = Column(Integer, nullable=False, default=0)
    status = Column(String(16), nullable=False, default="pending")
    validation_result = Column(JSONB, nullable=False, default=json_object)
    created_at = Column(DateTime(timezone=True), nullable=False, default=utc_now)
    finished_at = Column(DateTime(timezone=True))
    __table_args__ = (
        CheckConstraint("status IN ('pending','building','ready','failed')", name="ck_quant_dataset_status"),
        CheckConstraint("date_to >= date_from", name="ck_quant_dataset_dates"),
        Index("ix_quant_dataset_lookup", "market", "universe_id", "created_at"),
    )


class MarketRegimeSnapshot(Base):
    __tablename__ = "market_regime_snapshot"
    id = Column(BigInteger, primary_key=True)
    market = Column(String(8), nullable=False)
    trade_date = Column(Date, nullable=False)
    model_version = Column(String(64), nullable=False)
    regime = Column(String(16), nullable=False)
    market_score = Column(Float, nullable=False)
    max_equity_exposure = Column(Float, nullable=False)
    features = Column(JSONB, nullable=False, default=json_object)
    reasons = Column(JSONB, nullable=False, default=json_array)
    generated_at = Column(DateTime(timezone=True), nullable=False, default=utc_now)
    __table_args__ = (
        UniqueConstraint("market", "trade_date", "model_version", name="uix_market_regime_version"),
        CheckConstraint("regime IN ('risk_on','neutral','risk_off')", name="ck_market_regime_state"),
        Index("ix_market_regime_latest", "market", "trade_date"),
    )


class ModelDefinition(Base):
    __tablename__ = "model_definition"
    id = Column(Integer, primary_key=True)
    key = Column(String(80), nullable=False, unique=True)
    name = Column(String(160), nullable=False)
    model_type = Column(String(24), nullable=False)
    task_type = Column(String(24), nullable=False)
    frequency = Column(String(16), nullable=False)
    target_definition = Column(JSONB, nullable=False, default=json_object)
    enabled = Column(Boolean, nullable=False, default=True)
    default_config = Column(JSONB, nullable=False, default=json_object)
    supported_markets = Column(JSONB, nullable=False, default=json_array)
    created_at = Column(DateTime(timezone=True), nullable=False, default=utc_now)
    updated_at = Column(DateTime(timezone=True), nullable=False, default=utc_now, onupdate=utc_now)
    __table_args__ = (
        CheckConstraint(
            "model_type IN ('market_regime','time_series','cross_section','fusion')", name="ck_model_definition_type"
        ),
    )


class ModelRun(Base):
    __tablename__ = "model_run"
    id = Column(BigInteger, primary_key=True)
    uid = Column(Integer, ForeignKey("users.id", ondelete="RESTRICT"), nullable=False)
    task_id = Column(String(64))
    model_definition_id = Column(Integer, ForeignKey("model_definition.id", ondelete="RESTRICT"), nullable=False)
    model_key = Column(String(80), nullable=False)
    model_version = Column(String(96), nullable=False)
    run_type = Column(String(24), nullable=False)
    market = Column(String(8), nullable=False)
    universe_id = Column(Integer, ForeignKey("universe.id", ondelete="RESTRICT"), nullable=False)
    dataset_snapshot_id = Column(BigInteger, ForeignKey("quant_dataset_snapshot.id", ondelete="RESTRICT"))
    train_start = Column(Date)
    train_end = Column(Date)
    valid_start = Column(Date)
    valid_end = Column(Date)
    test_start = Column(Date)
    test_end = Column(Date)
    prediction_date = Column(Date)
    status = Column(String(20), nullable=False, default="draft")
    progress = Column(Integer, nullable=False, default=0)
    parameters = Column(JSONB, nullable=False, default=json_object)
    split_config = Column(JSONB, nullable=False, default=json_object)
    feature_config = Column(JSONB, nullable=False, default=json_object)
    target_config = Column(JSONB, nullable=False, default=json_object)
    metrics = Column(JSONB, nullable=False, default=json_object)
    feature_importance = Column(JSONB, nullable=False, default=json_object)
    artifact_uri = Column(Text)
    artifact_digest = Column(String(64))
    artifact_size = Column(BigInteger)
    code_commit = Column(String(64))
    warnings = Column(JSONB, nullable=False, default=json_array)
    error = Column(Text)
    created_at = Column(DateTime(timezone=True), nullable=False, default=utc_now)
    started_at = Column(DateTime(timezone=True))
    finished_at = Column(DateTime(timezone=True))
    __table_args__ = (
        CheckConstraint("run_type IN ('train','backtest','predict','walk_forward')", name="ck_model_run_type"),
        CheckConstraint(
            "status IN ('draft','training','candidate','production','retired','failed')", name="ck_model_run_status"
        ),
        CheckConstraint("progress BETWEEN 0 AND 100", name="ck_model_run_progress"),
        Index("ix_model_run_lookup", "market", "model_key", "status", "created_at"),
        Index(
            "uix_model_run_production_market_key",
            "market",
            "model_key",
            unique=True,
            postgresql_where=text("status = 'production'"),
        ),
    )


class ModelPublication(Base):
    __tablename__ = "model_publication"
    id = Column(BigInteger, primary_key=True)
    model_run_id = Column(BigInteger, ForeignKey("model_run.id", ondelete="RESTRICT"), nullable=False)
    previous_model_run_id = Column(BigInteger, ForeignKey("model_run.id", ondelete="SET NULL"))
    published_by = Column(Integer, ForeignKey("users.id", ondelete="RESTRICT"), nullable=False)
    reason = Column(Text, nullable=False)
    published_at = Column(DateTime(timezone=True), nullable=False, default=utc_now)


class ModelSignal(Base):
    __tablename__ = "model_signal"
    id = Column(BigInteger, primary_key=True)
    trade_date = Column(Date, nullable=False)
    instrument_id = Column(Integer, ForeignKey("instrument.id", ondelete="RESTRICT"), nullable=False)
    code = Column(String(32), nullable=False)
    market = Column(String(8), nullable=False)
    universe_id = Column(Integer, ForeignKey("universe.id", ondelete="RESTRICT"), nullable=False)
    model_version = Column(String(96), nullable=False)
    market_score = Column(Float)
    time_series_score = Column(Float)
    cross_section_score = Column(Float)
    risk_penalty = Column(Float, nullable=False, default=0)
    final_score = Column(Float, nullable=False)
    universe_rank = Column(Integer)
    predicted_return = Column(Float)
    signal = Column(String(16), nullable=False)
    reasons = Column(JSONB, nullable=False, default=json_array)
    score_components = Column(JSONB, nullable=False, default=json_object)
    generated_at = Column(DateTime(timezone=True), nullable=False, default=utc_now)
    __table_args__ = (
        UniqueConstraint(
            "market",
            "universe_id",
            "trade_date",
            "instrument_id",
            "model_version",
            name="uix_model_signal_market_universe",
        ),
        Index("ix_model_signal_ranking", "market", "universe_id", "trade_date", "universe_rank"),
    )


class PortfolioRecommendation(Base):
    __tablename__ = "portfolio_recommendation"
    id = Column(BigInteger, primary_key=True)
    trade_date = Column(Date, nullable=False)
    market = Column(String(8), nullable=False)
    universe_id = Column(Integer, ForeignKey("universe.id", ondelete="RESTRICT"), nullable=False)
    model_version = Column(String(96), nullable=False)
    market_regime_id = Column(BigInteger, ForeignKey("market_regime_snapshot.id", ondelete="RESTRICT"), nullable=False)
    status = Column(String(20), nullable=False, default="ready")
    max_equity_exposure = Column(Float, nullable=False)
    target_equity_exposure = Column(Float, nullable=False)
    config = Column(JSONB, nullable=False, default=json_object)
    summary = Column(JSONB, nullable=False, default=json_object)
    warnings = Column(JSONB, nullable=False, default=json_array)
    generated_at = Column(DateTime(timezone=True), nullable=False, default=utc_now)
    items = relationship("PortfolioRecommendationItem", cascade="all, delete-orphan", back_populates="recommendation")
    __table_args__ = (
        UniqueConstraint("trade_date", "universe_id", "model_version", name="uix_portfolio_recommendation"),
    )


class PortfolioRecommendationItem(Base):
    __tablename__ = "portfolio_recommendation_item"
    id = Column(BigInteger, primary_key=True)
    recommendation_id = Column(
        BigInteger, ForeignKey("portfolio_recommendation.id", ondelete="CASCADE"), nullable=False
    )
    instrument_id = Column(Integer, ForeignKey("instrument.id", ondelete="RESTRICT"), nullable=False)
    code = Column(String(32), nullable=False)
    rank = Column(Integer, nullable=False)
    target_weight = Column(Float, nullable=False, default=0)
    final_score = Column(Float, nullable=False)
    predicted_return = Column(Float)
    signal = Column(String(16), nullable=False)
    reasons = Column(JSONB, nullable=False, default=json_array)
    constraints = Column(JSONB, nullable=False, default=json_object)
    recommendation = relationship("PortfolioRecommendation", back_populates="items")
    __table_args__ = (
        UniqueConstraint("recommendation_id", "instrument_id", name="uix_portfolio_item"),
    )


QUANT_TABLES = (
    QuantDatasetSnapshot,
    MarketRegimeSnapshot,
    ModelDefinition,
    ModelRun,
    ModelPublication,
    ModelSignal,
    PortfolioRecommendation,
    PortfolioRecommendationItem,
)
